File: modules/screen-time-module/ios/ScreenTimeModule.swift
import ExpoModulesCore
import SwiftUI
import FamilyControls
import DeviceActivity
import ManagedSettings

public class ScreenTimeModule: Module {

    public func definition() -> ModuleDefinition {
        Name("ScreenTimeModule")

        Events("onSelectionChange")

        // MARK: - Authorization

        AsyncFunction("requestAuthorization") { () -> String in
            guard #available(iOS 16.0, *) else {
                return "denied:iOS 16+ required"
            }
            do {
                try await AuthorizationCenter.shared.requestAuthorization(for: .individual)
                return "approved"
            } catch {
                print("[ScreenTimeModule] Authorization error: \(error)")
                print("[ScreenTimeModule] Error localizedDescription: \(error.localizedDescription)")
                return "denied:\(error.localizedDescription)"
            }
        }

        AsyncFunction("getAuthorizationStatus") { () -> String in
            guard #available(iOS 16.0, *) else {
                return "notDetermined"
            }
            switch AuthorizationCenter.shared.authorizationStatus {
            case .approved:
                return "approved"
            case .denied:
                return "denied"
            case .notDetermined:
                return "notDetermined"
            @unknown default:
                return "notDetermined"
            }
        }

        // MARK: - App Picker

        AsyncFunction("showAppPicker") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            guard let self = self else { return }
            let store = SharedDataStore.shared
            await MainActor.run {
                guard let scene = UIApplication.shared.connectedScenes
                    .compactMap({ $0 as? UIWindowScene })
                    .first,
                      let rootVC = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController
                else { return }

                var topVC = rootVC
                while let presented = topVC.presentedViewController {
                    topVC = presented
                }

                let picker = ScreenTimeActivityPicker(store: store) { [weak self] in
                    topVC.dismiss(animated: true)
                    let sel = store.savedSelection
                    let count = sel.applicationTokens.count + sel.categoryTokens.count
                    self?.sendEvent("onSelectionChange", ["count": count])
                }
                let hostingController = UIHostingController(rootView: picker)
                hostingController.modalPresentationStyle = .pageSheet
                topVC.present(hostingController, animated: true)
            }
        }

        AsyncFunction("getSelectionCount") { () -> Int in
            guard #available(iOS 16.0, *) else { return 0 }
            let sel = SharedDataStore.shared.savedSelection
            return sel.applicationTokens.count + sel.categoryTokens.count
        }

        // MARK: - Block apps (no schedule)
        //
        // The model is now: pick apps → blocked. The only way to unlock is
        // to spend credits, which schedules a one-shot DeviceActivity event
        // to re-block at the unlock-window's intervalEnd.

        AsyncFunction("blockApps") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            self?.applyShieldsFromStore()
            self?.startAlwaysBlockSchedule()
            DeviceActivityCenter().stopMonitoring([.unlockWindow])
            SharedDataStore.shared.appsUnlocked = false
        }

        AsyncFunction("unblockAll") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            self?.removeAllShields()
            DeviceActivityCenter().stopMonitoring()
            SharedDataStore.shared.appsUnlocked = false
        }

        // MARK: - Shield (immediate apply/remove)

        AsyncFunction("applyShieldNow") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            self?.applyShieldsFromStore()
        }

        AsyncFunction("removeShieldNow") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            self?.removeAllShields()
        }

        // MARK: - Unlock window scheduling
        //
        // When the user spends credits to unlock, JS calls scheduleUnlockExpiry
        // with the duration in minutes. iOS schedules a DeviceActivity event
        // whose intervalEnd is `now + minutes`. At that moment, the system wakes
        // BrainLockMonitor (a separate process) which re-applies shields from
        // SharedDataStore.savedSelection — works even if the host app is killed.

        AsyncFunction("scheduleUnlockExpiry") { (minutes: Int) -> Void in
            guard #available(iOS 16.0, *) else { return }
            let cal = Calendar.current
            let minutes = max(1, minutes)

            // Nudge intervalStart 2 seconds into the future so iOS never
            // considers it "already passed" by the time startMonitoring runs.
            let startDate = Date().addingTimeInterval(2)
            let endDate = startDate.addingTimeInterval(TimeInterval(minutes * 60))

            // Include year/month/day/hour/minute/second so the schedule is
            // anchored to today rather than interpreted as a recurring daily
            // time-of-day window.
            let anchorComps: Set<Calendar.Component> = [.year, .month, .day, .hour, .minute, .second]
            let startComps = cal.dateComponents(anchorComps, from: startDate)
            let endComps = cal.dateComponents(anchorComps, from: endDate)

            let schedule = DeviceActivitySchedule(
                intervalStart: startComps,
                intervalEnd: endComps,
                repeats: false
            )

            SharedDataStore.shared.unlockExpiresAt = endDate
            SharedDataStore.shared.appsUnlocked = true

            let center = DeviceActivityCenter()
            center.stopMonitoring([.unlockWindow])
            do {
                try center.startMonitoring(.unlockWindow, during: schedule)
                print("[ScreenTimeModule] unlockWindow monitoring started, expires at \(endDate)")
            } catch {
                print("[ScreenTimeModule] startMonitoring failed: \(error)")
            }
        }

        AsyncFunction("cancelUnlockExpiry") { () -> Void in
            guard #available(iOS 16.0, *) else { return }
            DeviceActivityCenter().stopMonitoring([.unlockWindow])
            SharedDataStore.shared.appsUnlocked = false
            SharedDataStore.shared.unlockExpiresAt = nil
        }

        // MARK: - Unlock state (read-only convenience for JS)

        AsyncFunction("setAppsUnlocked") { (unlocked: Bool) -> Void in
            guard #available(iOS 16.0, *) else { return }
            SharedDataStore.shared.appsUnlocked = unlocked
        }

        AsyncFunction("getAppsUnlocked") { () -> Bool in
            guard #available(iOS 16.0, *) else { return false }
            return SharedDataStore.shared.appsUnlocked
        }

        AsyncFunction("getUnlockExpiresAt") { () -> Double in
            guard #available(iOS 16.0, *) else { return 0 }
            return SharedDataStore.shared.unlockExpiresAt?.timeIntervalSince1970 ?? 0
        }

        // MARK: - Ensure blocking on app launch
        //
        // Called once on app foreground. Re-applies shields if the user has
        // selected apps and is not currently within an active unlock window.
        // This is a belt-and-braces fallback in case ManagedSettings was
        // somehow cleared (extremely rare; ManagedSettings persists across
        // launches by default).

        AsyncFunction("ensureBlocking") { [weak self] () -> Void in
            guard #available(iOS 16.0, *) else { return }
            guard let self = self else { return }
            let store = SharedDataStore.shared
            let sel = store.savedSelection
            let hasSelection = !sel.applicationTokens.isEmpty || !sel.categoryTokens.isEmpty
            guard hasSelection else { return }

            if store.appsUnlocked {
                print("[ScreenTimeModule] Active unlock window — leaving shields removed")
                return
            }

            print("[ScreenTimeModule] ensureBlocking: re-applying shields")
            self.applyShieldsFromStore()
            self.startAlwaysBlockSchedule()
        }
    }

    // MARK: - Private Helpers

    @available(iOS 16.0, *)
    private func applyShieldsFromStore() {
        let sel = SharedDataStore.shared.savedSelection
        let managedStore = ManagedSettingsStore()
        managedStore.shield.applications = sel.applicationTokens.isEmpty ? nil : sel.applicationTokens
        managedStore.shield.applicationCategories = sel.categoryTokens.isEmpty ? nil : .specific(sel.categoryTokens)
        managedStore.shield.webDomains = sel.webDomainTokens.isEmpty ? nil : sel.webDomainTokens
        print("[ScreenTimeModule] Shields applied: \(sel.applicationTokens.count) apps, \(sel.categoryTokens.count) categories")
    }

    @available(iOS 16.0, *)
    private func startAlwaysBlockSchedule() {
        let center = DeviceActivityCenter()
        // A repeating midnight-to-midnight schedule. The system fires
        // intervalDidStart on the monitor extension every day at 00:00,
        // including after device reboots, so shields get re-applied
        // without the user opening the app.
        let schedule = DeviceActivitySchedule(
            intervalStart: DateComponents(hour: 0, minute: 0),
            intervalEnd: DateComponents(hour: 23, minute: 59),
            repeats: true
        )
        // Don't stop-then-start — just start. If already monitoring this
        // activity, startMonitoring replaces the existing schedule.
        do {
            try center.startMonitoring(.alwaysBlock, during: schedule)
            print("[ScreenTimeModule] alwaysBlock schedule registered")
        } catch {
            print("[ScreenTimeModule] alwaysBlock schedule failed: \(error)")
        }
    }

    @available(iOS 16.0, *)
    private func removeAllShields() {
        let managedStore = ManagedSettingsStore()
        managedStore.shield.applications = nil
        managedStore.shield.applicationCategories = nil
        managedStore.shield.webDomains = nil
        print("[ScreenTimeModule] All shields removed")
    }
}

// MARK: - DeviceActivityName Extension

@available(iOS 16.0, *)
extension DeviceActivityName {
    static let unlockWindow = Self("unlockWindow")
    static let alwaysBlock = Self("alwaysBlock")
}

File: modules/screen-time-module/src/ScreenTimeModule.ts
import { requireNativeModule, EventEmitter, type Subscription } from 'expo-modules-core';
import { Platform } from 'react-native';

interface ScreenTimeModuleNative {
  requestAuthorization(): Promise<string>;
  getAuthorizationStatus(): Promise<string>;
  showAppPicker(): Promise<void>;
  getSelectionCount(): Promise<number>;
  blockApps(): Promise<void>;
  unblockAll(): Promise<void>;
  applyShieldNow(): Promise<void>;
  removeShieldNow(): Promise<void>;
  ensureBlocking(): Promise<void>;
  scheduleUnlockExpiry(minutes: number): Promise<void>;
  cancelUnlockExpiry(): Promise<void>;
  setAppsUnlocked(unlocked: boolean): Promise<void>;
  getAppsUnlocked(): Promise<boolean>;
  getUnlockExpiresAt(): Promise<number>;
}

const NativeModule: ScreenTimeModuleNative | null = (() => {
  if (Platform.OS !== 'ios') return null;
  try {
    return requireNativeModule('ScreenTimeModule');
  } catch (e) {
    console.warn('ScreenTimeModule native module not found. The app needs to be built natively.');
    return null;
  }
})();

const emitter = NativeModule ? new EventEmitter(NativeModule as any) : null;

export const ScreenTime = {
  /** Whether Screen Time APIs are available (iOS only) */
  isAvailable: Platform.OS === 'ios',

  /** Listen for selection changes from the native app picker. */
  addSelectionChangeListener(callback: (event: { count: number }) => void): Subscription | null {
    if (!emitter) return null;
    return emitter.addListener('onSelectionChange', callback);
  },

  /** Request FamilyControls authorization. Returns 'approved' or 'denied'. */
  async requestAuthorization(): Promise<'approved' | 'denied'> {
    if (!NativeModule) throw new Error('Screen Time is only available on iOS');
    const result = await NativeModule.requestAuthorization();
    return result as 'approved' | 'denied';
  },

  /** Get current authorization status: 'approved', 'denied', or 'notDetermined'. */
  async getAuthorizationStatus(): Promise<
    'approved' | 'denied' | 'notDetermined'
  > {
    if (!NativeModule) return 'notDetermined';
    const result = await NativeModule.getAuthorizationStatus();
    return result as 'approved' | 'denied' | 'notDetermined';
  },

  /** Present the native FamilyActivityPicker to select apps/categories. */
  async showAppPicker(): Promise<void> {
    if (!NativeModule) throw new Error('Screen Time is only available on iOS');
    await NativeModule.showAppPicker();
  },

  /** Get the number of selected apps + categories. */
  async getSelectionCount(): Promise<number> {
    if (!NativeModule) return 0;
    return await NativeModule.getSelectionCount();
  },

  /** Apply the shield to currently-selected apps. Idempotent. */
  async blockApps(): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.blockApps();
  },

  /** Lift the shield and cancel any active unlock window. */
  async unblockAll(): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.unblockAll();
  },

  /** Immediately apply shields to selected apps. */
  async applyShieldNow(): Promise<void> {
    if (!NativeModule) throw new Error('Screen Time is only available on iOS');
    await NativeModule.applyShieldNow();
  },

  /** Immediately remove all shields. */
  async removeShieldNow(): Promise<void> {
    if (!NativeModule) throw new Error('Screen Time is only available on iOS');
    await NativeModule.removeShieldNow();
  },

  /** Re-apply shields on app launch if a selection exists and no unlock window is active. */
  async ensureBlocking(): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.ensureBlocking();
  },

  /**
   * Schedule the DeviceActivityMonitor extension to fire at `now + minutes`.
   * When the interval ends, the extension re-applies the shield even if the
   * host app has been killed. Call this from JS right after `removeShieldNow`.
   */
  async scheduleUnlockExpiry(minutes: number): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.scheduleUnlockExpiry(minutes);
  },

  /** Cancel any pending unlock-window expiry (e.g. when re-blocking manually). */
  async cancelUnlockExpiry(): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.cancelUnlockExpiry();
  },

  /** Write unlock state to shared UserDefaults so native code + extension can read it. */
  async setAppsUnlocked(unlocked: boolean): Promise<void> {
    if (!NativeModule) return;
    await NativeModule.setAppsUnlocked(unlocked);
  },

  /** Read unlock state from shared UserDefaults. */
  async getAppsUnlocked(): Promise<boolean> {
    if (!NativeModule) return false;
    return await NativeModule.getAppsUnlocked();
  },

  /** Read unlock expiry timestamp (unix seconds, 0 if no active window). */
  async getUnlockExpiresAt(): Promise<number> {
    if (!NativeModule) return 0;
    return await NativeModule.getUnlockExpiresAt();
  },
};
